ops/pebble.py

```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
             'services': services,
         }
         self._request('POST', '/v1/signals', body=body)
+
+
+class MultipartLargeFileParser:
+    """A limited purpose multi-part parser backed by files for memory efficiency."""
+
+    def __init__(self, boundary: bytes):
+        self._response = None
+        self._files = {}
+
+        # RFC 2046 says that the boundary string needs to be preceded by a CRLF.
+        # Unfortunately, the request library's header parsing logic strips off one of
+        # these, so we'll prime the buffer with that missing sequence.
+        self._buffer = b'\r\n'
+
+        # Prepare the MIME multipart boundary line patterns.
+        # Note: RFC 2046 notes optional "linear whitespace" (e.g. [ \t]) after the boundary pattern
+        # and the optional "--" suffix.  Pebble doesn't use this, so to simplify implementation I
+        # am ignoring it.
+        self._boundary = "\r\n--{}\r\n".format(boundary).encode()
+        self._terminal_boundary = "\r\n--{}--\r\n".format(boundary).encode()
+
+        # State vars, as we may enter the feed() function multiple times.
+        self._headers = None
+        self._part_type = None
+
+    def __del__(self):
+        for file in self._files.values():
+            os.unlink(file.name)
+
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer += data
+        # self._run_parser_fsm()
+
+        header_terminator = b'\r\n\r\n'
+
+        while True:
+            # Try to read headers, if present.
+            if self._headers is None:
+                if self._buffer.startswith(self._boundary) and header_terminator in self._buffer:
+                    # Data is everything after the boundary, up to and including the double CRLF
+                    # of the headers.
+                    data_start = len(self._boundary)
+                    data_end = self._buffer.index(header_terminator) + len(header_terminator)
+                    data = self._buffer[data_start:data_end]
+
+                    self._headers = self._parse_headers(data)
+                    self._detect_invalid_content_disposition()
+                    self._part_type = self._prepare_for_part_body()
+
+                    # Advance the buffer to point at the data.
+                    self._buffer = self._buffer[data_end:]
+                elif self._buffer.startswith(self._terminal_boundary):
+                    # We've reached the terminal boundary; ignore any remaining data
+                    self.terminated = True
+                    break
+                else:
+                    # Insufficient data to continue
+                    break
+
+            if self._part_type == 'response':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    data = self._buffer[:next_boundary_index]
+                    self._response = json.loads(data.decode())
```

Comment:
`json.loads()` takes a str or bytes or bytearray, so I don't think you need `.decode()` here.

Reply:
Turns out that json in python 3.5 only takes `str`.  So this needs to stay like it is.